### PR TITLE
Fix ToPIL

### LINKED ISSUE #1638: ToPIL tranformer
**Is your feature request related to a problem? Please describe.**
Although MONAI transformations work with numpy, there are some torchvision transformations that works only with PIL images. However, there is not any transformer in MONAI to convert `torch.tensor` or `numpy.ndarray` to `PIL`.

Also since we have MONAI `TorchVisionD` which enable using torchvision transformation in MONAI workflow, this `ToPIL` is an obvious missing component.

**Describe the solution you'd like**
We need a `ToPIL` transformer that can simply transform  `torch.tensor` or `numpy.ndarray` to `PIL`.

### PR DESCRIPTION
**Pull request by behxyz**
Fixes #1638 

### Description
Add ToPIL and ToPILD along with the unittests

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/transforms/utility/array.py [hunk 1]
```diff
@@ -261,6 +266,22 @@ def __call__(self, img: Union[List, Tuple, np.ndarray, torch.Tensor]) -> np.ndar
         return np.ascontiguousarray(img)
 
 
+class ToPIL(Transform):
+    """
+    Converts the input image (in the form of NumPy array or PyTorch Tensor) to PIL image
+    """
+
+    def __call__(self, img: Union[np.ndarray, torch.Tensor]) -> PILImage.Image:
```
Comment (rijobro) [inline monai/transforms/utility/array.py:274]:
```suggestion
    def __call__(self, img: Union[np.ndarray, torch.Tensor, PILImage.Image]) -> PILImage.Image:
```

**Review by rijobro — APPROVED** [review]

File: monai/transforms/utility/array.py (hunk 1 above)
Comment (behxyz) [reply monai/transforms/utility/array.py:274]: @rijobro Good catch! thanks :)

**Comment by behxyz** [comment]
@Nic-Ma @wyli @rijobro 
min tests are failing because now we have transformers that depends on `Pillow` and it is not installed by minimum requirements. Should I add pillow to `requirements-min.txt` or do you have other idea?

**Comment by rijobro** [comment]
How about changing this: `PILImage, _ = optional_import("PIL.Image")` to this: `PILImage, has_pil = optional_import("PIL.Image")`.

And then this:

```python
if isinstance(img, PILImage.Image):
    return img
if isinstance(img, torch.Tensor):
    img = img.detach().cpu().numpy()
return PILImage.fromarray(img)
```
to this:
```python
if isinstance(img, PILImage.Image):
    return img
if not has_pil:
    raise RuntimeError("ToPIL requires pillow to be installed")
if isinstance(img, torch.Tensor):
    img = img.detach().cpu().numpy()
return PILImage.fromarray(img)
```

And then any tests can have the decorator added `@skipUnless(has_pil, "PIL required")`.

Is that what you're looking for?

**Comment by rijobro** [comment]
Your tests would also need to use `optional_import`

**Comment by wyli** [comment]
> @Nic-Ma @wyli @rijobro
> min tests are failing because now we have transformers that depends on `Pillow` and it is not installed by minimum requirements. Should I add pillow to `requirements-min.txt` or do you have other idea?

thanks please add the relevant tests to this list: https://github.com/Project-MONAI/MONAI/blob/master/tests/min_tests.py#L29 so that they'll be skipped in the minimal test pipeline

**Comment by behxyz** [comment]
Thank you @rijobro and @wyli!
I'll add the tests to be excluded but the problem with min.tests is not that. It fails before running any test.

```
Traceback (most recent call last):
38
  File "C:\hostedtoolcache\windows\Python\3.8.7\x64\lib\runpy.py", line 194, in _run_module_as_main
39
['monai.apps', 'monai.data', 'monai.engines', 'monai.handlers', 'monai.transforms', 'monai.visualize']
40
    return _run_code(code, main_globals, None,
41
  File "C:\hostedtoolcache\windows\Python\3.8.7\x64\lib\runpy.py", line 87, in _run_code
42
    exec(code, run_globals)
43
  File "D:\a\MONAI\MONAI\tests\min_tests.py", line 139, in <module>
44
    assert sorted(err_mod) == ["monai.engines", "monai.handlers"]
45
AssertionError
```

**Comment by behxyz** [comment]
It fails at load_submodule, where more submodules have error than ["monai.engines", "monai.handlers]
https://github.com/Project-MONAI/MONAI/blob/9ae3005abd2ff73c8f9b04871994bae11b600d2f/tests/min_tests.py#L135

**Comment by wyli** [comment]
> It fails at load_submodule, where more submodules have error than ["monai.engines", "monai.handlers]
> https://github.com/Project-MONAI/MONAI/blob/9ae3005abd2ff73c8f9b04871994bae11b600d2f/tests/min_tests.py#L135

yes, `optional_import` is needed for 'import PIL' in the newly added unit tests as @rijobro mentioned

**Comment by wyli** [comment]
> > It fails at load_submodule, where more submodules have error than ["monai.engines", "monai.handlers]
> > https://github.com/Project-MONAI/MONAI/blob/9ae3005abd2ff73c8f9b04871994bae11b600d2f/tests/min_tests.py#L135
> 
> yes, `optional_import` is needed for 'import PIL' in the newly added unit tests as @rijobro mentioned

sorry it's incorrect. perhaps it's because of this `PILImage.Image` type not defined in the min. env https://github.com/Project-MONAI/MONAI/pull/1648/files#diff-24a052cc80d92db9eeddbd616450f309e9b88d293f187f469b9f73dbf4cf85f9R260

**Comment by behxyz** [comment]
> > > It fails at load_submodule, where more submodules have error than ["monai.engines", "monai.handlers]
> > > https://github.com/Project-MONAI/MONAI/blob/9ae3005abd2ff73c8f9b04871994bae11b600d2f/tests/min_tests.py#L135
> > 
> > 
> > yes, `optional_import` is needed for 'import PIL' in the newly added unit tests as @rijobro mentioned
> 
> sorry it's incorrect. perhaps it's because of this `PILImage.Image` type, not defined in the min. env https://github.com/Project-MONAI/MONAI/pull/1648/files#diff-24a052cc80d92db9eeddbd616450f309e9b88d293f187f469b9f73dbf4cf85f9R260

@wyli, sure, but how it can solve the issue with `load_submodules`? We already have tests that does not have optional import of PIL and things work fine: https://github.com/Project-MONAI/MONAI/blob/9ae3005abd2ff73c8f9b04871994bae11b600d2f/tests/test_pil_reader.py#L18

**Comment by wyli** [comment]
> > > > It fails at load_submodule, where more submodules have error than ["monai.engines", "monai.handlers]
> > > > https://github.com/Project-MONAI/MONAI/blob/9ae3005abd2ff73c8f9b04871994bae11b600d2f/tests/min_tests.py#L135
> > > 
> > > 
> > > yes, `optional_import` is needed for 'import PIL' in the newly added unit tests as @rijobro mentioned
> > 
> > 
> > sorry it's incorrect. perhaps it's because of this `PILImage.Image` type, not defined in the min. env https://github.com/Project-MONAI/MONAI/pull/1648/files#diff-24a052cc80d92db9eeddbd616450f309e9b88d293f187f469b9f73dbf4cf85f9R260
> 
> @wyli, sure, but how it can solve the issue with `load_submodules`? We already have a test that does not have optional import of PIL and things work fine:
> 
> https://github.com/Project-MONAI/MONAI/blob/9ae3005abd2ff73c8f9b04871994bae11b600d2f/tests/test_pil_reader.py#L18

could remove all `PILImage.Image` type hints, add `# type: ignore` wherever needed

**Comment by behxyz** [comment]
> > > > > It fails at load_submodule, where more submodules have error than ["monai.engines", "monai.handlers]
> > > > > https://github.com/Project-MONAI/MONAI/blob/9ae3005abd2ff73c8f9b04871994bae11b600d2f/tests/min_tests.py#L135
> > > > 
> > > > 
> > > > yes, `optional_import` is needed for 'import PIL' in the newly added unit tests as @rijobro mentioned
> > > 
> > > 
> > > sorry it's incorrect. perhaps it's because of this `PILImage.Image` type, not defined in the min. env https://github.com/Project-MONAI/MONAI/pull/1648/files#diff-24a052cc80d92db9eeddbd616450f309e9b88d293f187f469b9f73dbf4cf85f9R260
> > 
> > 
> > @wyli, sure, but how it can solve the issue with `load_submodules`? We already have a test that does not have optional import of PIL and things work fine:
> > https://github.com/Project-MONAI/MONAI/blob/9ae3005abd2ff73c8f9b04871994bae11b600d2f/tests/test_pil_reader.py#L18
> 
> could remove all `PILImage.Image` type hints, add `# type: ignore` wherever needed
> 
> type hints with `PILImage` is fine, but `PILImage.Image` is not defined when pillow is not installed

Exactly, that's the problem. Does anyone know a better solution for this? which is not particular to PIL. It's a pity to remove type checking for this. @Nic-Ma @rijobro @wyli

**Comment by behxyz** [comment]
> > > > > > It fails at load_submodule, where more submodules have error than ["monai.engines", "monai.handlers]
> > > > > > https://github.com/Project-MONAI/MONAI/blob/9ae3005abd2ff73c8f9b04871994bae11b600d2f/tests/min_tests.py#L135
> > > > > 
> > > > > 
> > > > > yes, `optional_import` is needed for 'import PIL' in the newly added unit tests as @rijobro mentioned
> > > > 
> > > > 
> > > > sorry it's incorrect. perhaps it's because of this `PILImage.Image` type, not defined in the min. env https://github.com/Project-MONAI/MONAI/pull/1648/files#diff-24a052cc80d92db9eeddbd616450f309e9b88d293f187f469b9f73dbf4cf85f9R260
> > > 
> > > 
> > > @wyli, sure, but how it can solve the issue with `load_submodules`? We already have a test that does not have optional import of PIL and things work fine:
> > > https://github.com/Project-MONAI/MONAI/blob/9ae3005abd2ff73c8f9b04871994bae11b600d2f/tests/test_pil_reader.py#L18
> > 
> > 
> > could remove all `PILImage.Image` type hints, add `# type: ignore` wherever needed
> > type hints with `PILImage` is fine, but `PILImage.Image` is not defined when pillow is not installed
> 
> Exactly, that's the problem. Does anyone know a better solution for this? which is not particular to PIL. It's a pity to remove type checking for this. @Nic-Ma @rijobro @wyli

I fixed it by keeping type checking as follow:
``` python
if TYPE_CHECKING:
    from PIL.Image import Image as PILImageImage
    from PIL.Image import fromarray as PILImageFromArray

    has_pil = True
else:
    PILImageFromArray, has_pil = optional_import("PIL.Image", name="fromarray")
    PILImageImage, _ = optional_import("PIL.Image", name="Image")
```

**Pull request merged**